gnomad v3 big query updates

mike-w-wilson opened the pull request:
## Summary
This PR encompasses multiple changes to get gnomAD v3 into BigQuery. The repo is now split between v2 and v3. The major difference being the data retrieval and filtering of the v3 sparse MT. The majority of the functionality remains similar to v2. Output has changed slightly in the genotype field as the PGT field is now transformed to `pgt1_non_ref` and `pgt2_non_ref`.  The v3 `bq_export_gnomad.py` does not filter down to a max global AF. The bug preventing the filtering in v2 has been fixed.
This code has been tested and the import can be found here: https://console.cloud.google.com/bigquery?organizationId=548622027621&project=broad-mpg-gnomad&p=broad-mpg-gnomad&d=mwilson&page=dataset .

mike-w-wilson committed:
Split into v2 and v3

mike-w-wilson committed:
Updated export and utils for gnomad_v3

mike-w-wilson committed:
Fixed freq filter bug

mike-w-wilson committed:
Updated imports for v2 and minor formatting

mike-w-wilson committed:
Updated __init__ for proper importing

mike-w-wilson committed:
Added PGT1_non_ref and PGT2_non_ref and gnomad version to path

mike-w-wilson committed:
Updated paths to include versioning

mike-w-wilson committed:
Updated view descriptions

mike-w-wilson committed:
Removed hard coded test mt

lfrancioli commented on gnomad_bigquery/v3/bq_export_gnomad.py:
```diff
@@ -0,0 +1,301 @@
+    select_expr = hl.is_defined(vep[mt.row_key])
+
+    freq = annotations.freq.ht()
+    select_expr = select_expr & hl.is_defined(freq[mt.row_key])
```
not sure if this makes sense if we're not filtering for freq. Also, I think we should keep an option to filter on freq and just not use it.

lfrancioli commented on gnomad_bigquery/v3/bq_export_gnomad.py:
```diff
@@ -0,0 +1,301 @@
+    def format_freq(ht: hl.Table, subset: str):
+        def filter_freqs(x: hl.expr.ArrayExpression):
+            if export_all_sex_freq:
+                return ~x[0].contains("platform") & ~x[0].contains("downsampling")
```
These don't exist in v3.

lfrancioli reviewed (commented):
First, thanks for catching that bug and modifying the PGT field as we had discussed. When it comes to the rest of this PR, the v3 portion is essentially a copy of the v2 with a few small changes here and there. As such, I think this needs a rework to ensure that we reuse rather than copy the code. 
At first glance, I'd say that essentially everything but `get_gnomad_data_for_bq` applies to both. As such I feel like adding a `version` option to the scripts may be the easiest way to handle this? If this is impractical and you'd rather keep the v2/v3 subfolder structure, a ton of code needs to go into shared utils (not necessarily one file).

mike-w-wilson committed:
Removed version-split repos, added version arg

mike-w-wilson committed:
removed * imports from init

mike-w-wilson commented:
Thanks for the feedback @lfrancioli ! I've removed the versioned repos and adding a version arg to the modules that need them. I do not love how I handle the imports and versioning but it was the only way I could think of doing this since the the version resources are just different and obviously the exome and genome vs just genomes. I have tested this code as well in the same space as the old: https://console.cloud.google.com/bigquery?organizationId=548622027621&project=broad-mpg-gnomad&p=broad-mpg-gnomad&d=mwilson&t=genomes&page=table . I again tested this on the head of v3 however none of the transcript consequences where greater than 3_prime_utr_variant so I removed that filter for the test so youll see all transcript consequences in the data loaded into bigquery. Again thanks for reviewing!

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    :param version: Version of gnomAD
+    :param export_missing_genotypes: whether to export only missing GTs
+    :param output_dir: Directory where to write parquet file
+    :param max_freq: Mazimum AF to allow in export 
```
Mazimum -> Maximum (is this my typo? 😂 )

mike-w-wilson replied:
Haha I think this is on me!

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+        mt = get_gnomad_data(data_type, non_refs_only=True)
+        vep = v2_vep(data_type).ht()
+        freq = v2_freq(data_type).ht()
+    elif version == 3:
```
not sure it's essential, but we could check that `data_type == genomes` here

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+        vep = v3_vep.ht()
+        freq = v3_freq.ht()
+    else:
+         raise DataException(f"There is no version {version} of gnomAD. Please choose one from {VERSIONS}")
```
"There is no version {version} of gnomAD {data_type}" ?

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    max_freq: Optional[float] = None,
+    least_consequence: str = None,
+    variant_index: bool = True,
+    non_refs_only=True
```
I think we can probably remove this param?

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    there is a transcript consequence that is at least the least_consequence passed
+    :param data_type: exomes or genomes
+    :param version: Version of gnomAD
+    :param export_missing_genotypes: whether to export only missing GTs
```
I think we should also remove this -- the current code isn't v3 compatible and including these massively increased size. thoughts?

mike-w-wilson replied:
This makes sense, considering the dataset will just grow larger and there doesnt seem to be any demand for it.

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+            vep.vep.transcript_consequences.any(lambda x: x.biotype == "protein_coding")
+        )
+    vep = vep.persist()
+    logger.info(
```
How about changing this to debug since it triggers a pass through the data? If you feel like this is info we should get to the user, fine to leave as is of course!

mike-w-wilson replied:
Yes, that makes sense to me! I don't think the user needs to know, it has just been helpful during testing.

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+
+    mt = mt.filter_rows(select_expr)
+    ht = mt.entries()
+    logger.info(
```
Same as above

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    """
+    def format_freq(ht: hl.Table, subset: str, version: int):
+        def filter_freqs(x: hl.expr.ArrayExpression):
+            if version==2 and export_all_sex_freq:
```
I was wrong in my previous review -- this would work for v3 too I think. Apologies!

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+        lcr = lcr_intervals_37.ht()
+        decoy = decoy_intervals_37.ht()
+        segdup = seg_dup_intervals_37.ht() 
+    elif version == 3:
```
add `data_type == "genomes"` ?

mike-w-wilson replied:
For these, I check after the args are parsed that the user isnt requesting exomes with v3 but this makes sense if the method is used in another module for some reason.

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+        freq = v3_freq.ht()
+        lcr = lcr_intervals_38.ht()
+    else:
+         raise DataException(f"There is no version {version} of gnomAD. Please choose one from {VERSIONS}")
```
"There is no version {version} of gnomAD {data_type}." ?

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+         raise DataException(f"There is no version {version} of gnomAD. Please choose one from {VERSIONS}")
+    
+    ht = mt.rows()
+    ht = ht.select().add_index()  # TODO: Add interesting annotations
```
The TODO can probably go at this point :)

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    :param data_type: exomes or genomes
+    :param version: Version of gnomAD
+    :param export_all_sex_freq: Whether to export sex frequencies alongside overall frequencies
+    :param export_rf: Whether to add random forest annotations
```
Maybe update to something more generic like `export_filters` ? We'd want to add the VQSR filters for v3

mike-w-wilson replied:
This sounds good. Are the VQSR filters located here: `gs://gnomad/variant_qc/genomes_v3/vqsr_classic.split.ht` ?  I was going to use this function https://github.com/broadinstitute/gnomad_qc/blob/b452bca9313370658eab156563793bf651e56a0c/gnomad_qc/v3/resources/variant_qc.py#L22

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+    
+    if version==2:
+        ht = ht.annotate(decoy=hl.is_defined(decoy[ht.locus]), segdup=hl.is_defined(segdup[ht.locus]))
+        if add_rf:
```
we should add the v3 VQSR filter / metrics here

mike-w-wilson replied:
Are these here: `gs://gnomad/variant_qc/genomes_v3/vqsr_classic.split.ht` ?

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+        if args.export_metadata:
+            if version == 2 :
+                meta = get_gnomad_meta(data_type=data_type, full_meta=True)
+            elif version == 3: 
```
add ` data_type == "genomes"` ?

mike-w-wilson replied:
Since this is the main and arg parse accounts for when data_type="exomes" and version=3, I'm not sure its really necessary but I guess it cant hurt.

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+            elif version == 3: 
+                meta = v3_meta.ht()
+            else:
+                raise DataException(f"There is no version {version} of gnomAD. Please choose one from {VERSIONS}")
```
"There is no version {version} of gnomAD {data_type}.

lfrancioli commented on gnomad_bigquery/bq_utils.py:
```diff
@@ -1,13 +1,18 @@
 logger = logging.getLogger("bq")
 logger.setLevel(logging.INFO)
 
+VERSIONS = [2, 3]
```
`GNOMAD_VERSIONS`?

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -0,0 +1,290 @@
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--slack_channel', help='Slack channel to post results and notifications to.')
+    parser.add_argument('--version', help="gnomAD version", choices=[2,3], type=int, required=True)
```
`choices = VERSIONS` ?

mike-w-wilson replied:
good call!

lfrancioli commented on gnomad_bigquery/bq_import_gnomad.py:
```diff
@@ -1,40 +1,54 @@
     parser = argparse.ArgumentParser()
     parser.add_argument('--exomes', help='Will import exomes data. At least one of --exomes or --genomes is required.', action='store_true')
     parser.add_argument('--genomes', help='Will import genomes data. At least one of --exomes or --genomes is required.', action='store_true')
+    parser.add_argument('--version', help='Version of gnomAD to import.', choices=[2,3], type=int, required=True)
```
`choices = VERSIONS` ?

lfrancioli commented on gnomad_bigquery/create_bq_views.py:
```diff
@@ -0,0 +1,130 @@
+    parser.add_argument('--exomes', help='Will import exomes data. At least one of --exomes or --genomes is required.', action='store_true')
+    parser.add_argument('--genomes', help='Will import genomes data. At least one of --exomes or --genomes is required.', action='store_true')
+    parser.add_argument('--dataset', help='Dataset to create the table in. (default: gnomad)', default='gnomad')
+    parser.add_argument('--version', help="Version of gnomAD", choices=[2,3], type=int, required=True)
```
`choices = VERSIONS` ?

lfrancioli reviewed (commented):
Hey @mike-w-wilson, thanks a lot for this!
So, overall I still feel like the the separation between generic and version-specific isn't quite where I'd ideally want it. I think having distinct scripts for the different versions but sharing functions would be preferable but maybe it's overthinking at this point.... it's just that e.g. adding v4 at this point essentially requires going through a lot of functions, wondering where it differs and adding additional conditions. 
Anyhow, happy to discuss and/or review if you want and try to tackle this in a cleaner way. Otherwise, also happy to say "this is overthinking, let's just use this".

A few small specific comments. Nothing major, really just details!

mike-w-wilson committed:
Added vqsr filters and changed to debug

mike-w-wilson commented:
Back to you @lfrancioli ! Based on our conversation in the gnomAD check in, I've address your comments and moved ahead with "let's just use this"  and get data to analysts instead of making this more generic though this is on my list of things to do.

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -124,23 +118,23 @@ def export_genotypes(
     :param version: Version of gnomAD
     :param export_all_sex_freq: Whether to export sex frequencies alongside overall frequencies
-    :param export_rf: Whether to add random forest annotations
+    :param export_filters: Whether to add random forest annotations
```
The message here should be updated

lfrancioli commented on gnomad_bigquery/bq_export_gnomad.py:
```diff
@@ -168,23 +162,24 @@ def get_freq_struct(x: hl.expr.ArrayExpression):
         vep = v3_vep.ht()
         freq = v3_freq.ht()
         lcr = lcr_intervals_38.ht()
+        vqsr = get_filtering_model("vqsr_alleleSpecificTrans", split=True).ht()
```
Entirely my bad but I gave you the wrong file here! While this has the correct VQSR run, it's before we set the threshold we wanted and thus doesn't correspond to what's in the release!
The correct file is:
gs://gnomad/annotations/hail-0.2/ht/genomes_v3/filtering/vqsr_alleleSpecificTrans.finalized.split.ht

lfrancioli replied:
I think for now, it's probably just easiest to hard-code this file somewhere with a TODO to update when v3 variant QC is in `gnomad_qc`

lfrancioli reviewed (approved):
One minor and one major (my bad!) things to modify, but they're both trivial so I'm already approving this. If you can just make these changes, then let's merge and run!

Thanks a lot for the work :)

mike-w-wilson committed:
Fixed VQSR file and removed unused args
